contents/javascript/Javascript-modules.md

```diff
@@ -0,0 +1,209 @@
+
+A more in-depth explanation of CommonJS can be found in the [Modules chapter of NodeJS API documentation](https://nodejs.org/docs/latest/api/modules.html).
+
+### Module Pattern
+
+Using a technique in JavaScript called _[IIFE (Immediately Invoked Function Expression)](https://developer.mozilla.org/en-US/docs/Glossary/IIFE)_, JavaScript developers can create module by wrapping their code in an IIFE.
+
+<box>
+
+An IIFE (Immediately Invoked Function Expression) is a JavaScript function that runs as soon as it is defined. <br>
+**Syntax:** `(function() { statements })();`
+
+_Source: [MDN Glossary - IIFE](https://developer.mozilla.org/en-US/docs/Glossary/IIFE)_
+</box>
+
+The example below will replicate the same example in previous chapters using module pattern. `index.js` needs the function `sumOfVariable` from `anExampleModule.js`. To achieve this, we use the `<script>` tag in HTML to _import_ `anExampleModule.js` for `index.js` to use.
+
+```html
+<!-- index.html -->
+
+<script src="./anExampleModule.js" />
+<script src="./index.js" />
+```
+
+```js
+// anExampleModule.js
+
+var anExampleModule = (function() {
+    var variableOne = 1;
+    var variableTwo = 2;
+
+    var sumOfVariable = function() {
+        return variableOne + variableTwo;
+    }
+
+    // We return an object with functions or variables
+    // that we want to be exposed
+    return {
+        sumOfVariable: sumOfVariable
+    }
+})()
+```
+
+```js
+// index.js
+
+anExampleModule.sumOfVariable(); // 3
+```
+
+The result of importing the modules in the HTML will be as follow:
+```js
+// anExampleModule.js
+
+var anExampleModule = (function() {
+    var variableOne = 1;
+    var variableTwo = 2;
+
+    var sumOfVariable = function() {
+        return variableOne + variableTwo;
+    }
+
+    // We return an object with functions or variables
+    // that we want to be exposed
+    return {
+        sumOfVariable: sumOfVariable
+    }
+})()
+
+// index.js
+
+anExampleModule.sumOfVariable(); // 3
+```
+
+A more in-depth explanation of module pattern can be found in the [this course blog on mastering module pattern](https://ultimatecourses.com/blog/mastering-the-module-pattern).
+
+## Which to use?
+
+With the launch of ES6 modules, one might think that why should we choose which one to use? Why not use ES6 modules right away?
+
+If an application does not allow you to use transpiler and bundler perhaps because of the additional overhead or you just need to create a simple and lean application, you might want to consider implementing the module pattern.
+
+If you will be developing a NodeJS application, you might want to use CommonJS because NodeJS does not natively support ES6 modules.
+
+All in all, the best implementation depends on the nature of your project.
```

Comment:
Slight rephrasing (to make it more concise):

----

## Which to use?
Although ES6 modules is the 'official' way to implement modules, there are situations where you might have to use one of the other options. Here are some examples:
* If your application does not allow you to use a transpiler and a bundler perhaps for some reason (e.g., because of the additional overhead they add), you can use the module pattern.
* If your application is NodeJS based, you might want to use CommonJS because NodeJS does not natively support ES6 modules.